### PR TITLE
feat(Quest): Add BreadcrumbForQuestId field to quest_template_addon

### LINKED ISSUE #8854: hownling fjord dont give a quest
### Current Behaviour

I can't archive "I've toured the fjord" for 1 quest. I've reached 104/105, but no one gives me any more quests. The quests are:
ID: 11411 Camp Winterhoof
ID: 12182 To Venomspite!
ID: 11286 The Artifacts of Steel Gate
ID: 11297 Keeping Watch on the Interlopers
ID: 11478 Outpost Over Yonder...
None of these quests are given to me.
I can't do LORE MASTER FOR 1 quest.

there is bug because "ID: 11297 Keeping Watch on the Interlopers" need a requisite "Suppressing the Elements" but this quest is completted and nps dont give me 11297
same situation for "The Artifacts of Steel Gate" the required is complited but nps dont give me another quest.

the nps dont give me nothing quest is impossible continued archivement

Could you just give me the archive? I'm only missing 1 mission.

### Expected Behaviour

Could you just give me the archive? I'm only missing 1 mission.

### Source

.

### Steps to reproduce the problem

.

### Extra Notes

_No response_

### Customizations

_No response_

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/a4ea6d60a24d1f499aaf8ca62d3f6b1ac857f823

### Operating system

Ubuntu 22.04

### Modules

- [mod-1v1-arena](https://github.com/azerothcore/mod-1v1-arena)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-anticheat](https://github.com/azerothcore/mod-anticheat)
- [mod-arena-3v3-solo-queue](https://github.com/azerothcore/mod-arena-3v3-solo-queue)
- [mod-arena-replay](https://github.com/azerothcore/mod-arena-replay)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-breaking-news-override](https://github.com/azerothcore/mod-breaking-news-override)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-chromiecraft-smartstone](https://github.com/chromiecraft/mod-chromiecraft-smartstone)
- [mod-costumes](https://github.com/azerothcore/mod-costumes)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-dmf-switch](https://github.com/azerothcore/mod-dmf-switch)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna](https://github.com/azerothcore/mod-eluna)
- [mod-evade-logging](https://github.com/heyitsbench/mod-evade-logging)
- [mod-instanced-worldbosses](https://github.com/nyeriah/mod-instanced-worldbosses)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-npc-spectator](https://github.com/gozzim/mod-npc-spectator)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-quests](https://github.com/azerothcore/mod-pvp-quests)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-transmog](https://github.com/azerothcore/mod-transmog)
- [mod-weekend-xp](https://github.com/azerothcore/mod-weekend-xp)
- [mod-zone-difficulty](https://github.com/azerothcore/mod-zone-difficulty)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-eventScripts](https://github.com/55Honey/Acore_eventScripts)
- [lua-ExchangeNpc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-RecruitAFriend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-SendAndBind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-TempAnnouncements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-Zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Server

ChromieCraft

### LINKED ISSUE #23813: Feature: Adding breadcrumb option to quest_template_addon as quest_template_addon.breadcrumb
### Describe your feature request or suggestion in detail

(assuming the quest_template_addon as any other %_addon are not gathered from sniffs or DBCs)

Since breadcrumb is always related to a quest, having as part of the core and quest_template_addon could reduce conditions clutter.

Quest A (breadcrumb) goes to Quest B in Classic Era live servers, you're only able to pick either Quest A or Quest B, similiar but no equally to how ExclusiveGroup is because in this case, you can't only complete either of them and be done for.

### Describe a possible solution to your feature or suggestion in detail

Let's use the Westbrook Garrison quest as an example.

-- Westbrook Garrison Needs Help! [239]
BreadCrumb = 11; // Riverpaw Gnoll Bounty

-- Riverpaw Gnoll Bounty [11]
BreadCrumb = 0; // No one breadcrumb, default to 0.

The quest that has Breadcrumb filled itself is the quest that's the breadcrumb towards the quest refered

So if you pick either quest 239 or quest 11 you only have 1 active at the time. And if you complete the quest targeted in the Breadcrumb, then the breadcrumb quest no longer can be picked.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by blinkysc**
## Changes Proposed:
This PR adds a new `BreadcrumbForQuestId` field to `quest_template_addon` to simplify breadcrumb quest management. When set, it indicates that the current quest is a breadcrumb leading to another quest, and both become mutually exclusive automatically.

**Behavior when Quest A has `BreadcrumbForQuestId = Quest B`:**
- Quest A (breadcrumb) becomes unavailable if Quest B is taken, complete, or rewarded
- Quest B (target) becomes unavailable if Quest A is in progress or complete (but NOT if rewarded - turning in the breadcrumb unlocks the target)

**Example:**
```sql
UPDATE quest_template_addon SET BreadcrumbForQuestId = 11286 WHERE ID = 11287;
```
This single line replaces 4 condition entries.

- [x] Core (units, players, creatures, game systems).
- [ ] Scripts (bosses, spell scripts, creature scripts).
- [x] Database (SAI, creatures, etc).

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/23813
- Closes https://github.com/chromiecraft/chromiecraft/issues/8854

## SOURCE:
The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [x] The changes promoted by this pull request come partially or entirely from another project (cherry-pick).
https://github.com/TrinityCore/TrinityCore/commit/5ed77113b637c49e99a610b00f385c51da67c3bf
## Tests Performed:
This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:
- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Apply SQL migration to add the column
2. Set a breadcrumb relationship: `UPDATE quest_template_addon SET BreadcrumbForQuestId = 11286 WHERE ID = 11287;`
3. Clear quest state: `.quest remove 11286` and `.quest remove 11287`
4. Complete prerequisite: `.quest complete 11261` then `.quest reward 11261`
5. Go to Sage Mistwalker (`.go creature id 24186`) - Quest 11286 (The Artifacts of Steel Gate) should be available
6. Accept 11286, go to Greatmother Ankha (`.go creature id 24135`) - Quest 11287 (Find Sage Mistwalker) should NOT be available
7. Abandon 11286 - Quest 11287 should reappear
8. Accept 11287, go to Sage Mistwalker - Quest 11286 should NOT be available
9. Turn in 11287 - Quest 11286 should now be available

## Known Issues and TODO List:
- [ ] Additional breadcrumb quests can be migrated to use this field once approved

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by blinkysc** [comment]
For some reason I thought Kinzunu authored this in Trinity... Sigh... I updated correct co-author

File: data/sql/base/db_world/quest_template_addon.sql [hunk 1]
```diff
@@ -38,6 +38,7 @@ CREATE TABLE `quest_template_addon` (
   `RequiredMaxRepValue` int NOT NULL DEFAULT '0',
   `ProvidedItemCount` tinyint unsigned NOT NULL DEFAULT '0',
   `SpecialFlags` int unsigned NOT NULL DEFAULT '0',
+  `BreadcrumbForQuestId` mediumint unsigned NOT NULL DEFAULT '0',
```
Comment (sogladev) [inline data/sql/base/db_world/quest_template_addon.sql]:
```suggestion
```
`data/sql/updates/pending_db_world/rev_1765373936949009445.sql` should handle creating this new field?

File: data/sql/updates/pending_db_world/rev_1765373936949009445.sql [hunk 2]
```diff
@@ -0,0 +1,287 @@
+--
+-- Add BreadcrumbForQuestId column to quest_template_addon
+--
+
+ALTER TABLE `quest_template_addon`
+ADD COLUMN `BreadcrumbForQuestId` mediumint unsigned NOT NULL DEFAULT '0' AFTER `SpecialFlags`;
```
Comment (sogladev) [inline data/sql/updates/pending_db_world/rev_1765373936949009445.sql]:
```suggestion
ADD COLUMN `BreadcrumbForQuestId` mediumint unsigned NOT NULL DEFAULT '0' AFTER `ExclusiveGroup`;
```

Could we group these related fields for readability?

e.g.,
```sql
-- Yorus Barleybrew
UPDATE `quest_template_addon` SET `NextQuestID` = 0, `ExclusiveGroup` = 0, `BreadcrumbForQuestId` = 1699
```

**Comment by TheSCREWEDSoftware** [comment]
Don't forget to add the co-authors of the cherry pick

**Comment by blinkysc** [comment]
> Don't forget to add the co-authors of the cherry pick

Yeah both are added. Not sure why it doesn't really show the 3 icons? But if you look at the commits it has co-authors

**Comment by TheSCREWEDSoftware** [comment]
It looks like some SQL needs fixing

> Quest 10374 is a breadcrumb quest but also has NextQuestID 5096 set
Quest 5093 is a breadcrumb quest but also has NextQuestID 5096 set
Quest 5094 is a breadcrumb quest but also has NextQuestID 5096 set
Quest 5095 is a breadcrumb quest but also has NextQuestID 5096 set
Quest 11287 is a breadcrumb quest but also has ExclusiveGroup 11287 set

**Comment by Gultask** [comment]
This deprecates https://github.com/azerothcore/azerothcore-wotlk/pull/24077?

**Comment by TheSCREWEDSoftware** [comment]
> I would suggest since you've the PR already open for Breadcrumb, having the SQL here (that fixes the current issue (wrong exclusivegroup) + proper breadcrumb set in your PR

So I assume yes, @blinkysc your input :)?

**Comment by blinkysc** [comment]
> This deprecates #24077?

Yeah it would, was so many of these breadcrumb quests showing up lol might as well implement this

File: src/server/game/Quests/QuestDef.h [hunk 3]
```diff
@@ -401,6 +402,7 @@ class Quest
     uint32 StartItemCount         = 0;
     uint32 RewardMailSenderEntry  = 0;
     uint32 SpecialFlags           = 0; // custom flags, not sniffed/WDB
+    uint32 BreadcrumbForQuestId   = 0;
```
Comment (Kitzunu) [inline src/server/game/Quests/QuestDef.h]: nitpick but this should be after exclusive group

File: data/sql/updates/pending_db_world/rev_1765373936949009445.sql [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+-- Alliance Rogue level 10 quests
+UPDATE `quest_template_addon` SET `PrevQuestId` = 0, `ExclusiveGroup` = 2206 WHERE `ID` IN (2206, 2238, 2242);
+UPDATE `quest_template_addon` SET `NextQuestId` = 0, `ExclusiveGroup` = 0 WHERE `ID` IN (2218, 2205, 2241);
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 2238 WHERE `ID` = 2218;
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 2206 WHERE `ID` = 2205;
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 2242 WHERE `ID` = 2241;
+
+-- Find Sage Mistwalker (breadcrumb for The Artifacts of Steel Gate)
+UPDATE `quest_template_addon` SET `ExclusiveGroup` = 0, `BreadcrumbForQuestId` = 11286 WHERE `ID` = 11287;
```
Comment (Gultask) [inline data/sql/updates/pending_db_world/rev_1765373936949009445.sql:278]:
```suggestion
-- Find Sage Mistwalker (breadcrumb for The Artifacts of Steel Gate)
UPDATE `quest_template_addon` SET `PrevQuestID` = 0, `ExclusiveGroup` = 0 WHERE (`ID` = 11286);
UPDATE `quest_template_addon` SET `ExclusiveGroup` = 0, `BreadcrumbForQuestId` = 11286 WHERE `ID` = 11287;
```

**Comment by TheSCREWEDSoftware** [comment]
Should close: https://github.com/chromiecraft/chromiecraft/issues/8854

File: src/server/game/Quests/QuestDef.h (hunk 3 above)
Comment (blinkysc) [reply src/server/game/Quests/QuestDef.h]: Done

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 5]
```diff
@@ -1217,6 +1218,37 @@ bool Player::SatisfyQuestExclusiveGroup(Quest const* qInfo, bool msg) const
...
+bool Player::SatisfyQuestBreadcrumb(Quest const* qInfo, bool msg) const
+{
+    uint32 breadcrumbForQuestId = qInfo->GetBreadcrumbForQuestId();
+    if (breadcrumbForQuestId)
+    {
+        QuestStatus status = GetQuestStatus(breadcrumbForQuestId);
+        if (status != QUEST_STATUS_NONE || IsQuestRewarded(breadcrumbForQuestId))
+        {
+            if (msg)
+                SendCanTakeQuestResponse(INVALIDREASON_DONT_HAVE_REQ);
```
Comment (sudlud) [inline src/server/game/Entities/Player/PlayerQuest.cpp:1230]: I prefer an empty line after these single line ifs

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 6]
```diff
@@ -1217,6 +1218,37 @@ bool Player::SatisfyQuestExclusiveGroup(Quest const* qInfo, bool msg) const
...
+
+    if (std::vector<uint32> const* breadcrumbs = sObjectMgr->GetBreadcrumbsForQuest(qInfo->GetQuestId()))
+    {
+        for (uint32 breadcrumbQuestId : *breadcrumbs)
+        {
+            QuestStatus status = GetQuestStatus(breadcrumbQuestId);
+            if (status == QUEST_STATUS_INCOMPLETE || status == QUEST_STATUS_COMPLETE || status == QUEST_STATUS_FAILED)
+            {
+                if (msg)
+                    SendCanTakeQuestResponse(INVALIDREASON_DONT_HAVE_REQ);
```
Comment (sudlud) [inline src/server/game/Entities/Player/PlayerQuest.cpp:1244]: Same here

File: src/server/game/Globals/ObjectMgr.h [hunk 7]
```diff
@@ -1145,6 +1145,17 @@ class ObjectMgr
...
     ExclusiveQuestGroups mExclusiveQuestGroups;
 
+    typedef std::unordered_map<uint32, std::vector<uint32>> BreadcrumbQuestMap;
+    BreadcrumbQuestMap _breadcrumbsForQuest;
+
+    [[nodiscard]] std::vector<uint32> const* GetBreadcrumbsForQuest(uint32 questId) const
+    {
+        auto itr = _breadcrumbsForQuest.find(questId);
+        if (itr != _breadcrumbsForQuest.end())
+            return &itr->second;
```
Comment (sudlud) [inline src/server/game/Globals/ObjectMgr.h:1155]: And here

File: src/server/game/Entities/Player/PlayerQuest.cpp [hunk 8]
```diff
@@ -1217,6 +1218,37 @@ bool Player::SatisfyQuestExclusiveGroup(Quest const* qInfo, bool msg) const
     return true;
 }
 
+bool Player::SatisfyQuestBreadcrumb(Quest const* qInfo, bool msg) const
+{
+    uint32 breadcrumbForQuestId = qInfo->GetBreadcrumbForQuestId();
+    if (breadcrumbForQuestId)
+    {
+        QuestStatus status = GetQuestStatus(breadcrumbForQuestId);
+        if (status != QUEST_STATUS_NONE || IsQuestRewarded(breadcrumbForQuestId))
```
[the remark this answers is not in the document]
Comment (blinkysc) [reply src/server/game/Entities/Player/PlayerQuest.cpp:1227]: This catches the edge case where a repeatable quest was rewarded but GetQuestStatus returns QUEST_STATUS_NONE. The IsQuestRewarded check is not redundant

File: data/sql/updates/pending_db_world/rev_1765373936949009445.sql [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+-- Yuka Screwspigot
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 4136 WHERE `ID` = 4324;
+
+-- Help Watcher Biggs
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 1396 WHERE `ID` = 9609;
+
+-- Taking a Stand
+UPDATE `quest_template_addon` SET `BreadcrumbForQuestId` = 12503 WHERE `ID` = 12795;
+UPDATE `quest_template_addon` SET `PrevQuestID` = 0 WHERE `ID` IN (12503, 12596);
```
[the remark this answers is not in the document]
Comment (blinkysc) [reply data/sql/updates/pending_db_world/rev_1765373936949009445.sql:231]: naw

**Comment by holmes111** [comment]
Can please someone review this PR, so Loremaster is obtainable again?
Thank you!

**Comment by TheSCREWEDSoftware** [comment]
https://github.com/chromiecraft/chromiecraft/issues/8536

Doesn't get closed but it's an issue that can be fixed here.

Currently `11501` is required for `11157`

11501 is a breadcrumb for 11157

**Review by Nyeriah — APPROVED** [review]
